### PR TITLE
impl(GCS+gRPC): examples for async `WriteObject()`

### PR DESCRIPTION
**Pull request by coryan**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13333)

### CODE REVIEW

File: google/cloud/storage/examples/storage_async_samples.cc [hunk 1]
```diff
@@ -200,17 +202,116 @@ void ReadObjectWithOptions(
...
+void WriteObject(google::cloud::storage_experimental::AsyncClient&,
+                 std::vector<std::string> const&) {
+  std::cerr << "AsyncClient::WriteObject() example requires coroutines\n";
+}
+
+void WriteObjectWithRetry(google::cloud::storage_experimental::AsyncClient&,
+                          std::vector<std::string> const&) {
   std::cerr
-      << "AsyncClient::ReadObjectStreaming() example requires coroutines\n";
+      << "AsyncClient::WriteObjectWithRetry() example requires coroutines\n";
```
Comment (dbolduc) [inline google/cloud/storage/examples/storage_async_samples.cc:314]: nit: `AsyncClient::WriteObjectWithRetry()` is not a thing. do you want just `WriteObject()` ?

File: google/cloud/storage/examples/storage_async_samples.cc [hunk 2]
```diff
@@ -200,17 +202,116 @@ void ReadObjectWithOptions(
...
+         std::istream& is) -> google::cloud::future<google::cloud::Status> {
+    while (token.valid()) {
+      std::vector<char> buffer(1024 * 1024);
+      is.read(buffer.data(), buffer.size());
+      buffer.resize(is.gcount());
+      auto w = co_await writer.Write(std::move(token),
+                                     gcs_ex::WritePayload(std::move(buffer)));
+      if (!w) co_return std::move(w).status();
+      token = *std::move(w);
+      if (is.eof()) break;
```
Comment (dbolduc) [inline google/cloud/storage/examples/storage_async_samples.cc:254]: should we do `while(token.valid() && !is.eof())` ?

File: google/cloud/storage/examples/storage_async_samples.cc [hunk 3]
```diff
@@ -200,17 +202,116 @@ void ReadObjectWithOptions(
...
+      if (absl::holds_alternative<gcs::ObjectMetadata>(state)) {
+        co_return absl::get<gcs::ObjectMetadata>(std::move(state));
+      }
+      // Refresh the upload id and reset the input source to the right offset.
+      upload_id = gcs::UseResumableUploadSession(writer.UploadId());
+      is.seekg(absl::get<std::int64_t>(std::move(state)));
+      auto status = co_await attempt(writer, token, is);
+      if (!status.ok()) continue;  // Error in upload, try again
+      auto f = co_await writer.Finalize(std::move(token));
+      if (f) co_return *std::move(f);  // Successfully uploaded, yay!
```
Comment (dbolduc) [inline google/cloud/storage/examples/storage_async_samples.cc:284]: Love the positivity in our samples.

File: google/cloud/storage/examples/storage_async_samples.cc [hunk 4]
```diff
@@ -200,17 +202,116 @@ void ReadObjectWithOptions(
...
+                                                        std::move(object_name)))
+                               .value();
+    while (token.valid()) {
+      std::vector<char> buffer(1024 * 1024);
+      is.read(buffer.data(), buffer.size());
+      buffer.resize(is.gcount());
+      token = (co_await writer.Write(std::move(token),
+                                     gcs_ex::WritePayload(std::move(buffer))))
+                  .value();
+      if (is.eof()) break;
```
Comment (dbolduc) [inline google/cloud/storage/examples/storage_async_samples.cc:225]: same comment about making this a loop condition.

File: google/cloud/storage/async/client.h [hunk 5]
```diff
@@ -252,6 +252,66 @@ class AsyncClient {
          std::move(options)});
   }
 
+  /**
+   * Uploads a new object without buffering.
+   *
+   * Use this API to upload objects without any client-side buffering. The API
+   * permits resuming an upload after a transient error. If the application
+   * can checkpoint a simple string the API can even be used to resume an upload
+   * even after the application restarts.
```
Comment (dbolduc) [inline google/cloud/storage/async/client.h:261]: If the application can checkpoint a simple string, the API can be used to resume an upload, even after the application restarts.

File: google/cloud/storage/async/client.h [hunk 6]
```diff
@@ -252,6 +252,66 @@ class AsyncClient {
...
+   *
+   * In both cases this API assumes the application can start sending data from
+   * an arbitrary point in the data source. That makes this API best suited to
+   * upload persistent data sources, such as files, of arbitrary size.
+   *
+   * To upload relatively small objects consider using `InsertObject` instead.
+   * To upload streaming data sources, where rewinding to an arbitrary point
+   * may be impossible consider using `StreamingUpload()`.
```
Comment (dbolduc) [inline google/cloud/storage/async/client.h:269]: s/impossible/impossible,/

File: google/cloud/storage/async/client.h [hunk 7]
```diff
@@ -252,6 +252,66 @@ class AsyncClient {
...
+   *
+   * This function always uses [resumable uploads][resumable-link]. The
+   * application can provide a `#RestoreResumableUploadSession()` option to
+   * resume a previously created upload. The returned object has accessors to
+   * query the session id and the next byte expected by GCS.
```
Comment (dbolduc) [inline google/cloud/storage/async/client.h:274]: nit: here and once below, do we capitalize "ID" ?

File: google/cloud/storage/async/client.h [hunk 8]
```diff
@@ -252,6 +252,66 @@ class AsyncClient {
...
+   *
+   * @note When resuming uploads it is the application's responsibility to save
+   *     the session id to restart the upload later. Likewise, it is the
+   *     application's responsibility to query the next expected byte and send
+   *     the remaining data without gaps or duplications.
+   *
+   * If the application does not provide a `#RestoreResumableUploadSession()`
+   * option, or it provides the `#NewResumableUploadSession()` option then this
+   * function will create a new resumable upload session is created.
```
Comment (dbolduc) [inline google/cloud/storage/async/client.h:283]: s/ is created//

Comment (dbolduc) [reply google/cloud/storage/async/client.h:283]: and consider `s/ then/, then/` on the line above.

File: google/cloud/storage/examples/storage_async_samples.cc [hunk 9]
```diff
@@ -326,13 +441,33 @@ void AutoRun(std::vector<std::string> const& argv) {
                                    std::to_string(metadata->generation())});
   }
 
+  if (!examples::UsingEmulator()) {
+    std::cout << "Creating file for upload [1]" << std::endl;
```
Comment (dbolduc) [inline google/cloud/storage/examples/storage_async_samples.cc:445]: [1] is ominous. More are coming?

**Review by dbolduc — APPROVED** [review]

File: google/cloud/storage/examples/storage_async_samples.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/examples/storage_async_samples.cc]: Done

File: google/cloud/storage/examples/storage_async_samples.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/examples/storage_async_samples.cc]: It is tricky.  The value is not defined after just calling the `std::ifstream` constructor. Using a `seekg()` clears the bit and makes the code clearer.  Let's do that.

File: google/cloud/storage/examples/storage_async_samples.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/storage/examples/storage_async_samples.cc]: Fixed.

File: google/cloud/storage/examples/storage_async_samples.cc (hunk 4 above)
Comment (coryan) [reply google/cloud/storage/examples/storage_async_samples.cc]: Fixed.

File: google/cloud/storage/async/client.h (hunk 5 above)
Comment (coryan) [reply google/cloud/storage/async/client.h]: Done

File: google/cloud/storage/async/client.h (hunk 6 above)
Comment (coryan) [reply google/cloud/storage/async/client.h]: Fixed,

File: google/cloud/storage/async/client.h (hunk 7 above)
Comment (coryan) [reply google/cloud/storage/async/client.h]: 🤷

File: google/cloud/storage/async/client.h (hunk 8 above)
Comment (coryan) [reply google/cloud/storage/async/client.h]: Fixed.

File: google/cloud/storage/async/client.h (hunk 7 above)
Comment (coryan) [reply google/cloud/storage/async/client.h]: and done.

File: google/cloud/storage/examples/storage_async_samples.cc (hunk 9 above)
Comment (coryan) [reply google/cloud/storage/examples/storage_async_samples.cc]: Maybe, but we can change it then.

File: google/cloud/storage/examples/storage_async_samples.cc (hunk 3 above)
Comment (dbolduc) [reply google/cloud/storage/examples/storage_async_samples.cc]: Boo

**Pull request merged**